wicker/core/datasets.py

```diff
@@ -151,55 +249,55 @@ def get_folder_size(bucket, prefix):
 
         return get_folder_size(bucket, key)
 
-    def _get_dataset_size(self):
-        """Gets total size of the dataset in bits
+    def _get_dataset_size(self, entire_dataset_parquet_dir_size: bool = False):
+        """Gets total size of the dataset in bits.
+
+        Args:
+            entire_dataset_parquet_dir_size (bool): Whether to get the entire parquet dir or just
+                the given partition dir size.
 
         Returns:
             int: total dataset size in bits
         """
+
         # intialize with size of parquet dir
-        par_dir_bytes = self._get_parquet_dir_size()
+        print("Parsing parquet and arrow dir for size.")
+        par_dir_bytes = self._get_parquet_dir_size(entire_dataset_parquet_dir_size=entire_dataset_parquet_dir_size)
 
         # need to know which columns are heavy pntr columns we go to for
-        # byte adding
+        # byte adding, parse which are heavy pointers based off metadata
+        # baked into schema
         schema = self.schema()
         heavy_pointer_cols = []
         for col_name in schema.get_all_column_names():
-            if schema.get_column(col_name).is_heavy_pointer:
+            column = schema.get_column(col_name)
+            if column is not None and column.is_heavy_pointer:
                 heavy_pointer_cols.append(col_name)
 
-        # Each individual row only knows which column file it goes to, so we have to
-        # neccesarily parse all rows :( to get the column files. This should be cached
-        # as metadata but that would require re-curating the datasets.
+        # create arrow table for parsing
         print("Creating arrow table")
         arrow_table = self.arrow_table()
 
         buckets_keys = set()
-        worker = ShuffleWorker(storage=self._storage)
+
         print("Processing through heavy pointers")
         for heavy_pntr_col in heavy_pointer_cols:
             print(f"Evaulating {heavy_pntr_col} for column file locations")
+            # Each individual row only knows which column file it goes to, so we have to
+            # neccesarily parse all rows :( to get the column files. This should be cached
+            # as metadata but that would require re-curating the datasets.
             for location_bytes in tqdm.tqdm(arrow_table[heavy_pntr_col].to_pylist()):
                 location = ColumnBytesFileLocationV1.from_bytes(location_bytes)
-                path = worker.s3_path_factory.get_column_concatenated_bytes_s3path_from_uuid(
+                path = self._s3_path_factory.get_column_concatenated_bytes_s3path_from_uuid(
                     location.file_id.bytes, dataset_name=self._dataset_id.name
                 )
                 bucket, key = path.replace("s3://", "").split("/", 1)
                 buckets_keys.add((bucket, key))
 
-        # we use existing shuffle worker here to grab the column file from
```

Comment:
FYI for Alborz that the way this old version of the code worked - it was declaring a `boto3.resource("s3")` for each bucket-key pair and using that with a Python `ThreadPool`. Both of these together are super inefficient to calculate the size of the dataset. The new way to calculate the size of the objects in S3 by splitting the S3 size requests among subprocesses (that have threads themselves) is much, much faster and correct.

Reply:
Thanks for finding this @convexquad , this is a nice improvement on top of the initial implementation @pickles-bread-and-butter had, I always wondered why it should be so slow, but we never got the time to dive deep into it.